test/util/fake.go

```diff
@@ -146,45 +167,54 @@ func (n *nominator) add(pi *framework.PodInfo, nominatingInfo *framework.Nominat
 		nodeName = pi.Pod.Status.NominatedNodeName
 	}
 
-	if n.podLister != nil {
-		// If the pod is not alive, don't contain it.
-		if _, err := n.podLister.Pods(pi.Pod.Namespace).Get(pi.Pod.Name); err != nil {
-			klog.V(4).InfoS("Pod doesn't exist in podLister, aborting adding it to the nominator", "pod", klog.KObj(pi.Pod))
+	if npm.podLister != nil {
+		// If the pod was removed or if it was already scheduled, don't nominate it.
+		updatedPod, err := npm.podLister.Pods(pi.Pod.Namespace).Get(pi.Pod.Name)
+		if err != nil {
+			logger.V(4).Info("Pod doesn't exist in podLister, aborted adding it to the nominator", "pod", klog.KObj(pi.Pod))
+			return
+		}
+		if updatedPod.Spec.NodeName != "" {
+			logger.V(4).Info("Pod is already scheduled to a node, aborted adding it to the nominator", "pod", klog.KObj(pi.Pod), "node", updatedPod.Spec.NodeName)
 			return
 		}
 	}
 
-	n.nominatedPodToNode[pi.Pod.UID] = nodeName
-	for _, npi := range n.nominatedPods[nodeName] {
+	npm.nominatedPodToNode[pi.Pod.UID] = nodeName
+	for _, npi := range npm.nominatedPods[nodeName] {
 		if npi.Pod.UID == pi.Pod.UID {
-			klog.V(4).InfoS("Pod already exists in the nominator", "pod", klog.KObj(npi.Pod))
+			logger.V(4).Info("Pod already exists in the nominator", "pod", klog.KObj(npi.Pod))
 			return
 		}
 	}
-	n.nominatedPods[nodeName] = append(n.nominatedPods[nodeName], pi)
+	npm.nominatedPods[nodeName] = append(npm.nominatedPods[nodeName], pi)
 }
 
-func (n *nominator) delete(p *v1.Pod) {
-	nnn, ok := n.nominatedPodToNode[p.UID]
+func (npm *nominator) delete(p *v1.Pod) {
+	nnn, ok := npm.nominatedPodToNode[p.UID]
 	if !ok {
 		return
 	}
-	for i, np := range n.nominatedPods[nnn] {
+	for i, np := range npm.nominatedPods[nnn] {
 		if np.Pod.UID == p.UID {
-			n.nominatedPods[nnn] = append(n.nominatedPods[nnn][:i], n.nominatedPods[nnn][i+1:]...)
-			if len(n.nominatedPods[nnn]) == 0 {
-				delete(n.nominatedPods, nnn)
+			npm.nominatedPods[nnn] = append(npm.nominatedPods[nnn][:i], npm.nominatedPods[nnn][i+1:]...)
+			if len(npm.nominatedPods[nnn]) == 0 {
+				delete(npm.nominatedPods, nnn)
 			}
 			break
 		}
 	}
-	delete(n.nominatedPodToNode, p.UID)
+	delete(npm.nominatedPodToNode, p.UID)
 }
 
 // UpdateNominatedPod updates the <oldPod> with <newPod>.
-func (n *nominator) UpdateNominatedPod(oldPod *v1.Pod, newPodInfo *framework.PodInfo) {
-	n.Lock()
-	defer n.Unlock()
+func (npm *nominator) UpdateNominatedPod(logger klog.Logger, oldPod *v1.Pod, newPodInfo *framework.PodInfo) {
```

Comment:
Do we need to replace 'n' with 'npm'? I recall that 'npm' is an abbreviation for 'nominatePodMap '. or just copy from upstream？

Reply:
I didn't notice that - just copied from upstream.